pkg/querier/queryrange/limits_test.go

```diff
@@ -623,3 +623,61 @@ func Test_MaxQuerySize_MaxLookBackPeriod(t *testing.T) {
 		})
 	}
 }
+
+func TestAcquireWithTiming(t *testing.T) {
+
+	ctx := context.Background()
+	sem := NewSemaphoreWithTiming(2)
+
+	// Channel to collect waiting times
+	waitingTimes := make(chan struct {
+		GoroutineID int
+		WaitingTime int64
+	}, 3)
+
+	tryAcquire := func(n int64, goroutineID int) {
+		elapsed, err := sem.Acquire(ctx, n)
+		if err != nil {
+			t.Errorf("Expected no error, got %v", err)
+		}
+		waitingTimes <- struct {
+			GoroutineID int
+			WaitingTime int64
+		}{goroutineID, elapsed.Milliseconds()}
+
+		defer sem.sem.Release(n)
+
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	go tryAcquire(1, 1)
+	go tryAcquire(1, 2)
+
+	// Sleep briefly to allow the first two goroutines to start running
+	time.Sleep(5 * time.Millisecond)
+
+	go tryAcquire(1, 3)
+
+	// Collect and sort waiting times
+	var waitingDurations []struct {
+		GoroutineID int
+		WaitingTime int64
+	}
+	for i := 0; i < 3; i++ {
+		waitingDurations = append(waitingDurations, <-waitingTimes)
+	}
+	// Find and check the waiting time for the third goroutine
+	var waiting3 int64
+	for _, waiting := range waitingDurations {
+		if waiting.GoroutineID == 3 {
+			waiting3 = waiting.WaitingTime
+			break
+		}
+	}
+
+	// Check that the waiting time for the third request is at least 10 milliseconds
```

Comment:
If you're expecting it to be at least 10ms, express that in the test please

Reply:
You could use `require.GreaterOrEqual(t, waiting3, 10*time.Millisecond)`

Reply:
sorry this is a typo  should be larger than 0 milliseconds, actually it should be  waiting3>0  && waiting3 <=5

Reply:
have updated